src/SelfGravity/Poi_UserWorkBeforePoisson_GREP.cpp

```diff
@@ -78,20 +73,38 @@ void Poi_UserWorkBeforePoisson_GREP( const double Time, const int lv )
    Profile_t *Phi_FaLv_New = Phi_eff[ FaLv ][     Sg_FaLv ];
    Profile_t *Phi_FaLv_Old = Phi_eff[ FaLv ][ 1 - Sg_FaLv ];
 
-   h_GREP_Lv_Data_New     = Phi_Lv_New  ->Data;
-   h_GREP_FaLv_Data_New   = Phi_FaLv_New->Data;
-   h_GREP_FaLv_Data_Old   = Phi_FaLv_Old->Data;
-   h_GREP_Lv_Radius_New   = Phi_Lv_New  ->Radius;
-   h_GREP_FaLv_Radius_New = Phi_FaLv_New->Radius;
-   h_GREP_FaLv_Radius_Old = Phi_FaLv_Old->Radius;
-   h_GREP_Lv_NBin_New     = Phi_Lv_New  ->NBin;
-   h_GREP_FaLv_NBin_New   = Phi_FaLv_New->NBin;
-   h_GREP_FaLv_NBin_Old   = Phi_FaLv_Old->NBin;
+   for (int b=0; b<Phi_Lv_New->NBin; b++) {
+//    check if the number of bin exceeds EXT_POT_GREP_NAUX_MAX
+//    Phi_FaLv_New and Phi_FaLv_Old are skipped since they have been checked earlier
+      if ( Phi_Lv_New->NBin > EXT_POT_GREP_NAUX_MAX )
+         Aux_Error( ERROR_INFO, "Number of bins = %d > EXT_POT_GREP_NAUX_MAX for GREP at lv = %d and SaveSg = %d !!\n",
+                    Phi_Lv_New->NBin, Lv, Sg_Lv );
+
+      h_ExtPotGREP[b                          ] = (real) Phi_Lv_New   ->Data  [b];
+      h_ExtPotGREP[b +   EXT_POT_GREP_NAUX_MAX] = (real) Phi_Lv_New   ->Radius[b];
+   }
+
+   for (int b=0; b<Phi_FaLv_New->NBin; b++) {
+      h_ExtPotGREP[b + 2*EXT_POT_GREP_NAUX_MAX] = (real) Phi_FaLv_New->Data  [b];
+      h_ExtPotGREP[b + 3*EXT_POT_GREP_NAUX_MAX] = (real) Phi_FaLv_New->Radius[b];
+   }
+
+   for (int b=0; b<Phi_FaLv_Old->NBin; b++) {
+      h_ExtPotGREP[b + 4*EXT_POT_GREP_NAUX_MAX] = (real) Phi_FaLv_Old->Data  [b];
+      h_ExtPotGREP[b + 5*EXT_POT_GREP_NAUX_MAX] = (real) Phi_FaLv_Old->Radius[b];
+   }
+
+
+// assign the value of h_ExtPotGenePtr
+   for (int i=0; i<6; i++)   h_ExtPotGenePtr[i] = (real**) (h_ExtPotGREP + i*EXT_POT_GREP_NAUX_MAX);
```

Comment:
IIUC, we don't really need `h_ExtPotGREP` at all and can just use `h_ExtPotGenePtr` everywhere. This will also remove the data copy above. The only minor drawback is that in `ExtPot_PassData2GPU_GREP()` we will need to call `cudaMemcpy()` twice since data are not continuous anymore. What do you think?

Reply:
I forgot that `h_ExtPotGREP` is for type casting from `double` to `real`. Then let's stick with the original approach.